x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,338 @@
+//	oa.cfg.Connection.SetOIDCTokenGenID(oa.tokenGenID)
+//	return cred.AccessToken, nil
+//}
+
+// invalidateAccessToken invalidates the access token, if the force flag is set to true (which is
+// only on a Reauth call) or if the tokenGenID of the connection is greater than or equal to the
+// tokenGenID of the OIDCAuthenticator. It should never actually be greater than, but only equal,
+// but this is a safety check, since extra invalidation is only a performance impact, not a
+// correctness impact.
+func (oa *OIDCAuthenticator) invalidateAccessToken(conn driver.Connection) {
+	oa.mu.Lock()
+	defer oa.mu.Unlock()
+	tokenGenID := conn.OIDCTokenGenID()
+	// If the connection used in a Reauth is a new connection it will not have a correct tokenGenID,
+	// it will instead be set to 0. In the absence of information, the only safe thing to do is to
+	// invalidate the cached accessToken.
+	if tokenGenID == 0 || tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		conn.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context, cfg *Config) error {
+	oa.invalidateAccessToken(cfg.Connection)
+	return oa.Auth(ctx, cfg)
+}
+
+// Auth authenticates the connection.
+func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	conn := cfg.Connection
+
+	oa.mu.Lock()
+	cachedAccessToken := oa.accessToken
+	oa.mu.Unlock()
+
+	if cachedAccessToken != "" {
+		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
+			userName:    oa.userName,
+			accessToken: cachedAccessToken,
+		})
+		if err == nil {
+			return nil
+		}
+		// this seems like it could be incorrect since we could be inavlidating an access token that
+		// has already been replaced by a different auth attempt, but the TokenGenID will prevernt
+		// that from happening.
+		oa.invalidateAccessToken(conn)
+		time.Sleep(invalidateSleepTimeout)
+	}
+
+	if oa.OIDCHumanCallback != nil {
+		return oa.doAuthHuman(ctx, cfg, oa.OIDCHumanCallback)
+	}
+
+	// Handle user provided or automatic provider machine callback.
+	var machineCallback OIDCCallback
+	if oa.OIDCMachineCallback != nil {
+		machineCallback = oa.OIDCMachineCallback
+	} else {
+		machineCallback, err = oa.providerCallback()
+		if err != nil {
+			return fmt.Errorf("error getting built-in OIDC provider: %w", err)
+		}
+	}
+
+	if machineCallback != nil {
+		return oa.doAuthMachine(ctx, cfg, machineCallback)
+	}
+	return newAuthError("no OIDC callback provided", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthHuman(_ context.Context, _ *Config, _ OIDCCallback) error {
+	// TODO GODRIVER-3246: Implement OIDC human flow
+	// Println is for linter
+	fmt.Println("OIDC human flow not implemented yet", oa.idpInfo)
```

Comment:
Remove print statement.

```suggestion
```

Reply:
The linter fails if I remove the print ;)